pkg/controller/instance/instance_controller.go

```diff
@@ -97,63 +100,71 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 				// since its linking to a bad FV.
 				return false
 			}
-			// Identify plan to be executed by this change
+
+			// Identify plan to be executed by this change.
 			var planName string
-			var ok bool
+			var planFound bool
 			if old.Spec.FrameworkVersion != new.Spec.FrameworkVersion {
-				// Its an Upgrade!
-				_, ok = fv.Spec.Plans["upgrade"]
-				if !ok {
-					_, ok = fv.Spec.Plans["update"]
-					if !ok {
-						_, ok = fv.Spec.Plans["deploy"]
-						if !ok {
-							log.Println("InstanceController: Could not find any plan to use for upgrade")
-							return false
-						}
-						ok = true // TODO: Do we need this here?
-						planName = "deploy"
-					} else {
-						planName = "update"
+				// Its an upgrade!
+				names := []string{"upgrade", "update", "deploy"}
+				for _, n := range names {
+					if _, planFound = fv.Spec.Plans[n]; planFound {
+						planName = n
+						break
 					}
-				} else {
-					planName = "upgrade"
+				}
+				if !planFound {
+					log.Printf("InstanceController: Could not find any plan to use to upgrade instance %v", new.Name)
+					return false
 				}
 			} else if !reflect.DeepEqual(old.Spec, new.Spec) {
 				for k := range parameterDifference(old.Spec.Parameters, new.Spec.Parameters) {
-					// Find the right parameter in the FV
+					// Find the spec of the updated parameter.
+					paramFound := false
 					for _, param := range fv.Spec.Parameters {
 						if param.Name == k {
-							planName = param.Trigger
-							ok = true
+							paramFound = true
+
+							if param.Trigger != "" {
+								planName = param.Trigger
+								planFound = true
+							}
+							// TODO: what should we do if more than one parameter was updated?
 						}
 					}
-					if !ok {
-						log.Printf("InstanceController: Instance %v updated parameter %v, but parameter not found in FrameworkVersion %v\n", new.Name, k, fv.Name)
-					} else if planName == "" {
-						_, ok = fv.Spec.Plans["update"]
-						if !ok {
-							_, ok = fv.Spec.Plans["deploy"]
-							if !ok {
-								log.Println("InstanceController: Could not find any plan to use for update")
-							} else {
-								planName = "deploy"
+
+					if paramFound {
+						if !planFound {
+							// The parameter doesn't have a trigger, try to find the corresponding default plan.
+							names := []string{"update", "deploy"}
+							for _, n := range names {
+								if _, planFound = fv.Spec.Plans[n]; planFound {
+									planName = n
+									planFound = true
+									break
+								}
```

Comment:
Wait, previously (line 140 on the left) there is a default fallback `planName = "deploy"` if no plan was found. This is not the case on the right, correct?

Reply:
The original line #140 isn't setting `planName=deploy` as a default fallback — it's actually only executed if `ok == true`, which means that `deploy` was found.

Line #151 might be an attempt to use `deploy` as a default fallback plan, but if that line is ever executed, `ok` will be `false`, so the `PlanExecution` will never be created.

Reply:
You're right. But since we spent quite some time understanding all the `ifs and elses` of the logic,  I believe we need to extract this portion of code into separate methods and provide a few tests for it (similar to what I did for `parameterDifference`. Wdyt?

Reply:
To be honest I think that instead going further with the refactoring of the predicate function, I think we should get rid of it and move most of the logic to the `Reconcile` function, see my writeup on #422 for more details.

Instead of writing low level unit tests for something that we should change soon, I would rather start working on #422 or write a few unit tests to prepare the ground for #422.

Reply:
BTW it looks like the original intention was to create a `PlanExecution` for `deploy` even if no changes were detected.

@runyontr recently created a PR that includes a change to make this happen: https://github.com/kudobuilder/kudo/pull/367/files#r297890298